### ISSUE TITLE
Several questions regarding using ExecuteNetwork in ARM NN

### ISSUE THREAD

**Issue by HungYangChang** [ARM-software/armnn#578]
Hello Arm developer,

I have switched from Tensorflow lite to ARM NN and ARM-CL because it seems Tensorflow lite is unable to support  [layer-level splitting of task](https://discuss.tensorflow.org/t/does-tesnorflow-lite-supports-such-layer-level-splitting/3812). Therefore, I would like to use the **ExcuteNetwork** to run the BERT model (.tflite).

I have several questions:
Q1. To run the simulation with tflite on ARMnn, I used example code (available [here](https://drive.google.com/drive/folders/18EeS7Kh0EfWCbQK4pUz1QBrgHbqai_Sq?usp=sharing)):
`./build/tests/ExecuteNetwork  -m mobilenet_v1_0.25_224_quant.tflite -c CpuAcc --iterations=10 -f tflite-binary  -i input -o Reshape_1 -e 1`

I check my model with Netron to see input and output. (As shown below)
<img width="164" alt="Screen Shot 2021-09-10 at 09 16 06" src="https://user-images.githubusercontent.com/70759685/132859060-6d877ce3-f227-4c3a-be48-a4e5f3d968f0.png">
<img width="176" alt="Screen Shot 2021-09-10 at 09 16 13" src="https://user-images.githubusercontent.com/70759685/132859062-9519f2ee-209e-4120-83b3-8981d9e8c340.png">
I have referred to [#494](https://github.com/ARM-software/armnn/issues/494) and I used -i input -o as Reshape_1 but it seems incorrect to do so?  

```
Fatal: Armnn Error: No output binding found for subgraph:0 and name:Reshape_1. Possible outputs are: ['MobilenetV1/Predictions/Reshape_1' ]  at function GetNetworkOutputBindingInfo [/home/shunya/hungyang/armnn-dist/armnn/src/armnnTfLiteParser/TfLiteParser.cpp:4009]
Info: Shutdown time: 0.02 ms
```

Q2 I am trying to run on my own BERT.tflite model (available [here](https://drive.google.com/drive/folders/18EeS7Kh0EfWCbQK4pUz1QBrgHbqai_Sq?usp=sharing)):
`./build/tests/ExecuteNetwork -c CpuAcc -f tflite-binary -m ./bert-128.tflite -i input -o tanh`

However, it shows the following error when I tried to run on BERT.tfilte: 
It seems “ONE_HOT” is not a supported operation. (I have check [Arm NN Tf Lite Parser](https://arm-software.github.io/armnn/latest/parsers.xhtml#S6_tf_lite_parser)) Would it be supported soon or do I need to add it by myself?

```
Error: Failed to parse operator #2 within subgraph #0 error: Operator not supported. subgraph:0 operator:2 opcode_index:1 opcode:85 / ONE_HOT  at function ParseUnsupportedOperator [/home/shunya/hungyang/armnn-dist/armnn/src/armnnTfLiteParser/TfLiteParser.cpp:901]
Fatal: Armnn Error: Failed to parse operator #2 within subgraph #0 error: Operator not supported. subgraph:0 operator:2 opcode_index:1 opcode:85 / ONE_HOT  at function ParseUnsupportedOperator [/home/shunya/hungyang/armnn-dist/armnn/src/armnnTfLiteParser/TfLiteParser.cpp:901]
```
Note: the bert file is able to run with TensorFlow lite but it is not generated by ArmnnConverter.

Q3.  How exactly can we divide layers into the different cores? 
For example, say I have a 4 layers neural network and I wanna apply it on a 4-core CPU platform, How can I assign the first 2 layers into 3 CPU cores, the third layer into another CPU core? I know Armnn implements the workloads on top of ACL, the workload is divided into multiple smaller workloads and executed on different cores. I have referred to [this file](https://github.com/ARM-software/ComputeLibrary/blob/8b3fc248b2f0f1873852c97b69f669b5d77cf55e/arm_compute/runtime/CPP/CPPScheduler.h#L59), but I still don't know how exactly can we divide layers into the different cores.

Q4. I have asked, "Does ARM NN provide layer-wise granularity inference information?" in another [post](https://github.com/ARM-software/ComputeLibrary/issues/921#issuecomment-913520448)
I am wondering is it good enough to just use -e 1  to check layer-wise granularity inference information
Or should I follow what the post says and use tflite parser?

Operating System: Linux
Thanks for your answering and help in advance :)

===========================================================================================
More info about my setting:
ARM_NN_version = v21.05
arm_compute_version=v21.05 Build options: {'embed_kernels': '1', 'opencl': '1', 'neon': '1', 'debug': '0', 'validation_tests': '0', 'benchmark_tests': '1', 'extra_cxx_flags': '-fPIC', 'arch': 'arm64-v8a', 'benchmark_examples': '1'} Git hash=f7399fd616cc8ccedfae668da3023e589a7ef558
Platform: HiKey970

**Comment by HungYangChang** [comment]
I have solved the first problem by modifying  code into
`./build/tests/ExecuteNetwork -m mobilenet_v1_0.25_224_quant.tflite -c CpuAcc --iterations=10 -f tflite-binary -i input -o MobilenetV1/Predictions/Reshape_1`
![image](https://user-images.githubusercontent.com/70759685/132868762-38ee5210-4c9f-4b90-bb07-122058048196.png)

**Comment by Colm-in-Arm** [comment]
Hi @HungYangChang 

I'll skip Q1 as you've resolved it yourself.

Q2: ONE_HOT is not a supported operation. I have not see any plan to implement it. In fact, it is not an operator I'd heard of before. You could attempt to add it to the reference backend yourself. We would welcome the contribution. Then when executing specify -c CpuAcc,CpuRef so that layer will fall back to the reference implementation. 

Q3: Does Arm NN + ACL have a mechanism to tie layer execution to a NEON core? Short answer is no. You can play around with the graph and scheduler as they did [here](https://arxiv.org/pdf/1903.05898.pdf) but this is most definitely not trivial.

I'm sorry I don't have a more detailed answer for you - this stuff is way past my knowledge. You might be better off asking the question in [ComputeLibrary](https://github.com/ARM-software/ComputeLibrary)

Q4: Do you want information about the inference execution or do you want to guide which backend will execute a layer?

Specifying "-e 1" to execute network will enable the generation of json files that describe the execution. It will show which layers execute on which backends and much more besides.

The BackendSelectionHint is a mechanism to indicate that you want a backend to execute a specified layer. For example, if you had a model with 10 layers but you knew that 2 of those layers execute faster on the GPU and CPU you would pass CpuAcc,GpuAcc as backendPreferences to the Optimizer but then use BackendSelectionHint for those 2 layers. So long as GPU could support those 2 layers they will be executed on it.

Colm

**Comment by HungYangChang** [comment]
Hello @Colm-in-Arm and ARM NN developer,

First of all, thanks for your answer.

I have a few following questions regarding Excutenetwork:

1. For Q2, I have tried using MoblieBERT from [Tesnorflow Hub](https://tfhub.dev/tensorflow/lite-model/mobilebert/1/default/1), and was able to run the bert model. I guess now I will bypass ONE_HOT operation, I will add backend support if I still need that.
The command I use: 
`./build/tests/ExecuteNetwork -m ./mobilebert_1_default_1.tflite -f tflite-binary -c CpuRef -i input_ids,input_mask,segment_ids -o end_logits,start_logits --input-tensor-shape=1,384:1,384:1,384`


2. However, when I tried to run H-128_L-10_A-2_I-1024.tflite (available [here](https://drive.google.com/drive/folders/18EeS7Kh0EfWCbQK4pUz1QBrgHbqai_Sq?usp=sharing)) that I generate by myself.
The command I use: 

`taskset -c 0-8 ./build/tests/ExecuteNetwork  -m ./H-128_L-1_A-2_I-1024.tflite -f tflite-binary -c CpuRef -i input_mask:0,input_ids:0,segment_ids:0  -o start_logits:0,end_logits:0  --input-tensor-shape=1,64:1,64:1,64`

It shows the following error
```
Error: Failed to parse operator #8 within subgraph #0 error: Reshape operator target shape input buffer data is null:  at function ParseReshape [/home/shunya/hungyang/armnn-dist/armnn/src/armnnTfLiteParser/TfLiteParser.cpp:2373]
Fatal: Armnn Error: Failed to parse operator #8 within subgraph #0 error: Reshape operator target shape input buffer data is null:  at function ParseReshape [/home/shunya/hungyang/armnn-dist/armnn/src/armnnTfLiteParser/TfLiteParser.cpp:2373]
```

I have used [Netron ](https://github.com/lutzroeder/netron) to check both mobilebert_1_default_1.tflite and H-128_L-1_A-2_I-1024.tflite  but still don't know how to deal with this error, can you share any clue regrading this error.

![image](https://user-images.githubusercontent.com/70759685/133100334-0b88dceb-67fc-4ae5-aff0-9070549be59e.png)

**Comment by Colm-in-Arm** [comment]
Hi @HungYangChang 

A minor comment on the command line you are using for "mobilebert_1_default_1.tflite" there is no need to specify "--input-tensor-shape".

Looking at the error on the reshape operator: What's happening here is the parser is attempting to access the output shape data as the second input operand but the data is empty. The target shape has neither been specified as a builtin_options or as a parameter and the parser rejects it as a bad model. I cannot access that model on Google drive. If you expand the "shape" input does it have any data?

Colm.

**Comment by HungYangChang** [comment]
Sorry for the inconvenience.

Here is the [link ](https://drive.google.com/drive/folders/18EeS7Kh0EfWCbQK4pUz1QBrgHbqai_Sq?usp=sharing)to my file

> If you expand the "shape" input does it have any data?

May I kindly ask how to expand the "shape" input?  
Note: Now I am using --input-tensor-shape=1,64:1,64:1,64 (1 represent my batch size and 64 represent sequence length of my BERT model)

**Comment by Colm-in-Arm** [comment]
Hi @HungYangChang 

I'm sorry I wasn't clear in my earlier messages. I am not permitted to access a model unless I fully understand its origin. A reference to TensorFlow Hub is fine as it has a clear license and usage description.

In the previous image from Netron on the right hand side the inputs are listed. If you press the "+" character on the shape input.

COlm.

**Comment by HungYangChang** [comment]
Hello @Colm-in-Arm 


The input and output shape of Reshape operator (8) in my H-128_L-10_A-2_I-1024.tflite
![image](https://user-images.githubusercontent.com/70759685/133287542-b278b5ca-1060-4a42-bdd1-48c477f73b35.png)

Other info:
The input and output shape of my H-128_L-10_A-2_I-1024.tflite model
![image](https://user-images.githubusercontent.com/70759685/133287018-f1ed9d71-000e-4bdf-ae63-24a7aa12a4f9.png)

The input and output shape of the MoblieBERT from TensorFlow Hub
![image](https://user-images.githubusercontent.com/70759685/133287117-09fd0af3-2219-4908-9494-e3cc68fb095f.png)

Thanks again for your great help :)

**Comment by HungYangChang** [comment]
I list input and output of Pack operator and Dequantize which are right before Reshape operator
1. Pack operator 
![image](https://user-images.githubusercontent.com/70759685/133319221-77c61547-6466-45d0-8058-f3aff87a8dc0.png)
2. Dequantize
![image](https://user-images.githubusercontent.com/70759685/133319280-2d46112e-278d-46db-bacf-9cc280847743.png)

Hope this might be helpful :)

I also compare the Reshape operator in MoblieBERT from TensorFlow Hub. (which works correctly) but still cannot tell what incur the error in my H-128_L-10_A-2_I-1024.tflite model
![image](https://user-images.githubusercontent.com/70759685/133320313-18282e77-b4a6-40ea-b3a2-cdee1f31ef40.png)

[H-128_L-1_A-2_I-1024.tflite.zip](https://github.com/ARM-software/armnn/files/7164596/H-128_L-1_A-2_I-1024.tflite.zip)

**Comment by HungYangChang** [comment]
@Colm-in-Arm 

Problem solved:
The bert model I have in "H-128_L-10_A-2_I-1024.tflite " is the model with dynamic input size (i.e. batch size and sequence length) I tried to include the batch size and sequence length into my model and the compiler won't complain anymore.

Thanks for your help :)
